src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -34,98 +32,104 @@ public SocketInput(MemoryPool2 memory, IThreadPool threadPool)
-                    return new IncomingBuffer
-                    {
-                        Data = data,
-                        DataPtr = dataPtr,
-                    };
+                    _tail = _memory.Lease();
+                }
+
+                var iterator = new MemoryPoolIterator2(_tail, _tail.End);
+                iterator.CopyFrom(buffer, offset, count);
+
+                if (_head == null)
+                {
+                    _head = _tail;
                 }
-            }
 
-            _pinned = _memory.Lease(minimumSize);
-            return new IncomingBuffer
+                _tail = iterator.Block;
+            }
+            else
             {
-                Data = _pinned.Data,
-                DataPtr = _pinned.Pin() + _pinned.End
-            };
+                RemoteIntakeFin = true;
+            }
+
+            Complete();
         }
 
         public void IncomingComplete(int count, Exception error)
         {
-            Action awaitableState;
-
-            lock (_sync)
+            // Unpin may called without an earlier Pin 
+            if (_pinned != null)
             {
-                // Unpin may called without an earlier Pin 
-                if (_pinned != null)
+
+                _pinned.End += count;
+
+                if (_head == null)
                 {
-                    _pinned.Unpin();
-
-                    _pinned.End += count;
-                    if (_head == null)
-                    {
-                        _head = _tail = _pinned;
-                    }
-                    else if (_tail == _pinned)
-                    {
-                        // NO-OP: this was a read into unoccupied tail-space
-                    }
-                    else
-                    {
-                        _tail.Next = _pinned;
-                        _tail = _pinned;
-                    }
+                    _head = _tail = _pinned;
                 }
-                _pinned = null;
-
-                if (count == 0)
+                else if (_tail == _pinned)
                 {
-                    RemoteIntakeFin = true;
+                    // NO-OP: this was a read into unoccupied tail-space
                 }
-                if (error != null)
+                else
                 {
-                    _awaitableError = error;
+                    _tail.Next = _pinned;
+                    _tail = _pinned;
                 }
 
-                awaitableState = Interlocked.Exchange(
-                    ref _awaitableState,
-                    _awaitableIsCompleted);
+                _pinned = null;
+            }
 
-                _manualResetEvent.Set();
+            if (count == 0)
+            {
+                RemoteIntakeFin = true;
             }
+            if (error != null)
+            {
+                _awaitableError = error;
+            }
+
+            Complete();
+        }
+
+        public void AbortAwaiting()
```

Comment:
nit: move this back after `ConsumingComplete` so there's less impact on the diff.